enterprise/server/scheduling/scheduler_server/scheduler_server.go

```diff
@@ -1443,7 +1495,18 @@ func (s *SchedulerServer) LeaseTask(stream scpb.Scheduler_LeaseTaskServer) error
 			}
 		}
 
+		// If shutting down, generate a reconnect token. This indicates to the
```

Comment:
Instead of handing out this token at shut down time you could generate a "lease token" at the time the lease is obtained and use it in a similar manner. This matters if you plan to generalize this to non-shutdown cases since you won't have the opportunity to send a "reconnect" token in those cases.

Reply:
ah good point

Reply:
Done, except I kept the name `reconnect_token` for clarity (I don't think this will be useful for anything other than reconnecting the same lease?).

Also, we can no longer use `token != ""` as a signal that the server is shutting down, so I changed things so the server returns an explicit shutdown error instead, which the client checks for when `Recv()` fails.

Reply:
Slight modification to my last comment - the client no longer explicitly checks for the shutdown error, and now retries whenever it sees any `Unavailable` error. We now return an `Unavailable` error when the server is shutting down, and we might also see `Unavailable` in some other disconnect scenarios as well.